Get the Open Humans integration to a usable state

beaugunderson opened the pull request:
In addition to adding a hidden URL for a user to link and unlink their barcodes to Open Humans at `/authed/open-humans/` this PR also adds support for the `?next=/some/url/` feature for login redirection (I got tired of having to navigate somewhere from `/authed/portal/` every time I got logged out). :)

It looks like this:

![screenshot](https://s3.amazonaws.com/f.cl.ly/items/2p250x2M3r2o1V3M001a/Screen%20Shot%202015-01-14%20at%208.44.45%20PM.png)

cc @wasade, @madprime

wasade commented:
Thanks, @beaugunderson. Aiming to review todayish.

@squirrelo, can you comment on the logging out bit that @beaugunderson mentioned?

beaugunderson commented:
To explain the `next` feature more:

Prior to this pull request if you visit `/authed/some-url/` while logged out tornado redirects you to `/auth/login/?next=/authed/some-url/` by default. However, that `next` parameter was not used anywhere within the login form or the login handler, so it was being ignored--meaning no matter what URL you were trying to get to you get redirected to `/authed/portal/` after you log in.

I noticed this because I was doing local development, restarting the server, getting logged out, and having to navigate to `/authed/open-humans/` again and again. IIRC the logging out may happen because a session secret is generated at app startup? But regardless, this will actually be an important feature for Open Humans because we'll need to send people a link to `/authed/open-humans/`, and they'll probably be logged out when they try to visit--meaning they'll get redirected to `/authed/portal/` and have to enter the URL again.

This PR contains an implementation of the proper `next` redirect behavior which fixes all of that. :)

beaugunderson commented:
Re: my question above about whether the session secret is generated at startup, [it is](https://github.com/biocore/american-gut-web/blob/51071bb34466c9bf4f8aeeb695200ad075b5341a/amgut/webserver.py#L56). Meaning that sessions are lost across app restarts (OK-ish in production since restarts probably don't happen often, annoying in development since the server restarts when it detects any code has changed).

squirrelo commented:
@beaugunderson explained the `next` pretty well. Setting next is a feature built into tornado, but using next is not. Special, but this fix will take care of that.

wasade commented on amgut/templates/open-humans.html:
```diff
@@ -0,0 +1,79 @@
+    <p>Open Humans is an exciting project that puts participants in charge of
+      data sharing.</p>
+
+    <p>Members of Open Humans can aggregate data from various research studies
```
Possible that users may see this about concerned about sharing identifying details, possible to mention that identifying details are not shared?

madprime replied:
I recommend being more specific, since arguably a lot of things are potentially identifying, especially when combined with other information.

e.g.
This will export to Open Humans: Your raw sequencing files and survey answers.
This will not share: Your name, email address, or other confidential account information.

wasade commented:
Just one comment. Only performed static analysis so far. Hopefully can pull down pr and test today

beaugunderson commented:
That last commit just adds the `american-gut` scope to the requested scopes during the OAuth2 authorization.

@wasade Anything you need from me to move this forward?

wasade commented:
Have not been able to pull it down due to being tied up teaching. The earliest I can get to it is Feb 2. @adamrp @josenavas @squirrelo @ElDeveloper, are any of you able to pull down and test?

ElDeveloper commented:
I'm terribly unfamiliar with the current status of the project so I'm
not sure I would be the best person to review this PR at this time,
however I'm happy to do this with someone else.

On (Jan-25-15| 3:16), Daniel McDonald wrote:
> Have not been able to pull it down due to being tied up teaching. The earliest I can get to it is Feb 2. @adamrp @josenavas @squirrelo @ElDeveloper, are any of you able to pull down and test?
> 
> ---
> Reply to this email directly or view it on GitHub:
> https://github.com/biocore/american-gut-web/pull/361#issuecomment-71367132

beaugunderson commented:
@wasade Just checking in since you said you might have some time for this today--let me know if there's anything I can do to help. :)

wasade commented:
Was on vacation all last week, and terrible internet with limited to know free time the weeks prior. I can get to this later this week, unless @ElDeveloper @josenavas @squirrelo or @adamrp are able to get to it earlier

wasade commented:
@beaugunderson, how do we navigate to this page short of manually entering the URL? 

Currently getting the following when I go to `/authed/open-humans/`:

```bash
[E 150205 13:34:08 web:1228] Uncaught exception GET /authed/open-humans/ (::1)
    HTTPRequest(protocol='http', host='localhost:8884', method='GET', uri='/authed/open-humans/', version='HTTP/1.1', remote_ip='::1', headers={'Connection': 'keep-alive', 'Accept-Language': 'en-US,en;q=0.8', 'Accept-Encoding': 'gzip, deflate, sdch', 'Cookie': '_gat=1; skid=InRlc3Qi|1423168443|56f34ca14745431540e428a476efd89a599ef7f9; _ga=GA1.1.1304853201.1412276249', 'Host': 'localhost:8884', 'Accept': 'text/html,application/xhtml+xml,application/xml;q=0.9,image/webp,*/*;q=0.8', 'User-Agent': 'Mozilla/5.0 (Macintosh; Intel Mac OS X 10_8_5) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/40.0.2214.94 Safari/537.36'})
    Traceback (most recent call last):
      File "/Users/mcdonadt/.virtualenvs/python27/lib/python2.7/site-packages/tornado/web.py", line 1115, in _stack_context_handle_exception
        raise_exc_info((type, value, traceback))
      File "/Users/mcdonadt/.virtualenvs/python27/lib/python2.7/site-packages/tornado/web.py", line 1298, in wrapper
        result = method(self, *args, **kwargs)
      File "/Users/mcdonadt/rs/american-gut-web/amgut/handlers/open_humans.py", line 33, in get
        self.finish()
      File "/Users/mcdonadt/.virtualenvs/python27/lib/python2.7/site-packages/tornado/web.py", line 728, in finish
        raise RuntimeError("finish() called twice.  May be caused "
    RuntimeError: finish() called twice.  May be caused by using async operations without the @asynchronous decorator.
```

When I click on the export data button, I get this:

```bash
[E 150205 13:34:53 web:1228] Uncaught exception GET /authed/connect/open-humans/ (::1)
    HTTPRequest(protocol='http', host='localhost:8884', method='GET', uri='/authed/connect/open-humans/', version='HTTP/1.1', remote_ip='::1', headers={'Accept-Language': 'en-US,en;q=0.8', 'Accept-Encoding': 'gzip, deflate, sdch', 'Host': 'localhost:8884', 'Accept': 'text/html,application/xhtml+xml,application/xml;q=0.9,image/webp,*/*;q=0.8', 'User-Agent': 'Mozilla/5.0 (Macintosh; Intel Mac OS X 10_8_5) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/40.0.2214.94 Safari/537.36', 'Connection': 'keep-alive', 'Referer': 'http://localhost:8884/authed/open-humans/', 'Cookie': '_gat=1; skid=InRlc3Qi|1423168443|56f34ca14745431540e428a476efd89a599ef7f9; _ga=GA1.1.1304853201.1412276249'})
    Traceback (most recent call last):
      File "/Users/mcdonadt/.virtualenvs/python27/lib/python2.7/site-packages/tornado/web.py", line 1115, in _stack_context_handle_exception
        raise_exc_info((type, value, traceback))
      File "/Users/mcdonadt/.virtualenvs/python27/lib/python2.7/site-packages/tornado/web.py", line 1298, in wrapper
        result = method(self, *args, **kwargs)
      File "/Users/mcdonadt/rs/american-gut-web/amgut/handlers/open_humans.py", line 101, in get
        self.authorize_redirect(
    AttributeError: 'OpenHumansLoginHandler' object has no attribute 'authorize_redirect'
[E 150205 13:34:53 web:1635] 500 GET /authed/connect/open-humans/ (::1) 25.45ms
```

beaugunderson commented:
As to the first question, there's no way aside from visiting the URL manually right now (this is because we're doing a "soft launch" and didn't want everyone to have export access without us sending an email first).

As to the exceptions, I haven't seen either of those in my testing so I'll look ASAP!

beaugunderson commented:
Just to be sure, can you verify which commit you're running against? The `finish() called twice` did happen but that was early on in development, it was later fixed. :)

wasade commented:
I did a pull of the branch, so it should be latest. Can double check in 30
On Feb 5, 2015 2:48 PM, "Beau Gunderson" <notifications@github.com> wrote:

> Just to be sure, can you verify which commit you're running against? The finish()
> called twice did happen but that was early on in development, it was
> later fixed. :)
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/biocore/american-gut-web/pull/361#issuecomment-73136144>
> .
>

wasade commented:
Am @ c2861f3, so latest. Sorry for not confirming that last night

beaugunderson commented:
@wasade OK, I've reproduced the first error (I was confused because it doesn't affect functionality and the page still loads, but it does show up in the logs). Diving into that one now.

wasade commented:
Awesome, good to know I wasn't just insane :smile:

beaugunderson commented:
@wasade OK, fixed the first and I think the second was caused by you not having `open_humans_tornado_oauth2` installed in your virtualenv--I added a warning message if that's the case. It's in `install_requires` so running setup should get it for you. :)

cc: @madprime

wasade commented:
Okay, things seem cleared up. But now when I click the export button, it just hangs. No debug output, but chrome is indicating "Connecting...". Expected?

wasade commented:
More detail, got a connection time out with this URL:

`https://openhumans.org/oauth2/authorize/?scope=read+write+american-gut&redirect_uri=%2Fauthed%2Fconnect%2Fopen-humans%2F&client_id=`

beaugunderson commented:
Ah, hmm. I think that is why I added the `self.finish()` that I just took out for being "redundant"--it seems like one code path works with it and one without it! Will investigate.

beaugunderson commented:
@wasade Oh, re-reading your comment now that I'm not on my phone, `https://openhumans.org/` is the default for the authorize URL but that URL is not in production yet! Not actually related to the async stuff.

You'll want to set OPEN_HUMANS_BASE_URL like so in your config file for the testing/staging environment:

```
OPEN_HUMANS_BASE_URL = https://staging.openhumans.org
```

For the initial limited production release it will be:

```
OPEN_HUMANS_BASE_URL = https://stealth.openhumans.org
```

And then once we release to the public the default URL (https://openhumans.org) will be fine.

We've been using keybase.io to send encrypted API keys/secrets, I just sent you an invite there (it's a super easy wrapper around GPG but if you want to transfer another way that's also fine with me).

Once that's set up I'll send you the API key/secret for staging and production. :sparkles:

wasade commented:
Sounds good. I'm up on keybase.io, username is mcdonadt

wasade commented:
Get redirected now to a login on open humans, expected? Seeeeeems correct :)

beaugunderson commented:
@wasade OK, API keys/secret sent!

And yes, that's the correct behavior. :)

beaugunderson commented:
@wasade Any idea when this will make it to production? We'll likely have to delay our launch if it's not in by Monday but of course would be happy with some additional breathing room! Let me know if there's anything I can do to help the process. :)

wasade commented:
@beaugunderson, I'd like to do it this week. Will keep you posted. Thanks again for the contributions!

madprime commented:
@wasade - thanks! This week -- ASAP, really -- would be much appreciated.

We have to start circulating press materials next week. If we don't have our import features all operational at that point, that could likely mean we push everything back a week.

That would be embarrassing - the alternative is we somehow de-emphasize American Gut in our launch, which is also very unpalatable. I'd much rather not be in that situation. :)
